### PR TITLE
Cache result of getBestFitRule (fixes #575)

### LINKED ISSUE #575: property-updates-report is slow with many dependencies & rules
Not really sure why this happens, but when using a version rules file with the `property-updates-report` target, the report generation is very slow. I tried to implement a workaround in #365 (and added a reproducer today), since I wasn't able to track down the root cause of the issue... If someone has a better idea, I would be open to another approach.

### PR DESCRIPTION
**Pull request by TobiX**
This was identified as a performance hotspot and the result is static
for each run, adding a cache dramatically speeds up some executions.

The included IT shows the problem: Without this optimization, it runs
over a minute, with the fix it finishes instantly.

(Reopened from #365 because I did stupid things with our fork)

### CODE REVIEW

File: src/it/it-property-updates-report-002-slow/pom.xml [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      </dependency>
+    </dependencies>
+  </dependencyManagement>
+
+  <build>
+    <pluginManagement>
+      <plugins>
+        <plugin>
+          <artifactId>maven-site-plugin</artifactId>
+          <version>3.7</version>
```
Comment (slawekjaranowski) [inline src/it/it-property-updates-report-002-slow/pom.xml:189]: try @maven-site-plugin.version@  - property defined in parent pom

File: src/it/it-property-updates-report-002-slow/pom.xml [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+  <reporting>
+    <plugins>
+      <plugin>
+        <artifactId>maven-project-info-reports-plugin</artifactId>
+        <version>2.1</version>
+        <reportSets>
+          <reportSet/>
+        </reportSets>
+      </plugin>
```
Comment (slawekjaranowski) [inline src/it/it-property-updates-report-002-slow/pom.xml:203]: we can add: `<excludeDefaults>true</..>` than this will be not needed.

**Review by slawekjaranowski — COMMENTED** [review]
If we have another IT test which execute: `property-updates-report`

I would consider to remove old, after add moving assertions to this test. 
One test will be enough.

**Comment by slawekjaranowski** [comment]
There is: https://github.com/mojohaus/versions-maven-plugin/tree/master/src/it/it-property-updates-report-001 
can be merged to one test.

Pleas check build status 😄

**Comment by TobiX** [comment]
Hmm, I'm used to write one test per issue, but I can certainly extend the existing test to also show this problem...

**Comment by slawekjaranowski** [comment]
IT tests take long time - I try to minimize duplicate test.
If you think that both tests do something else - ok, we can have its.

**Comment by TobiX** [comment]
I tried to combine both ITs, unfortunately the validation script for the existing test is quite sensitive to dependency changes, so for the sake of stability I'd rather not combine both tests...

**Comment by slawekjaranowski** [comment]
I see a warning in IT build ...

```
[INFO] Building: it-property-updates-report-002-slow/pom.xml
[INFO] artifact org.junit.jupiter:junit-jupiter: checking for updates from central
[INFO] artifact org.junit.jupiter:junit-jupiter-engine: checking for updates from central
Warning:  Could not validate integrity of download from https://repo.maven.apache.org/maven2/org/junit/jupiter/junit-jupiter-engine/maven-metadata.xml.sha1
org.eclipse.aether.transfer.ChecksumFailureException: Checksum validation failed, no checksums available
...
[INFO] artifact org.junit.jupiter.junit-jupiter-engine:junit-jupiter-engine: checking for updates from central
Warning:  /org/junit/jupiter/junit-jupiter-engine/maven-metadata.xml.sha1
java.lang.NullPointerException
    at org.codehaus.mojo.mrm.impl.digest.SHA1DigestFileEntry.<init> (SHA1DigestFileEntry.java:67)
    at org.codehaus.mojo.mrm.impl.digest.SHA1DigestFileEntry$Factory.create (SHA1DigestFileEntry.java:148)
``` 

maybe it is a bug in mrm plugin ... but I would not introduce such warning

can you a litel investigate it?

**Comment by TobiX** [comment]
I just dropped that particular dependency from the test... It's still plenty slow before my change to fail the test before my fix...

**Comment by slawekjaranowski** [comment]
@TobiX - Thanks

**Pull request merged**